substrate/frame/multisig/src/migrations/v2/mod.rs

```diff
@@ -0,0 +1,155 @@
+struct LazyMigrationV2<T: Config, W: weights::WeightInfo>(PhantomData<(T, W)>);
+impl<T: Config, W: weights::WeightInfo> SteppedMigration for LazyMigrationV2<T, W> {
+	type Cursor = (T::AccountId, [u8; 32]);
+	// Without the explicit length here the construction of the ID would not be infallible.
+	type Identifier = MigrationId<15>;
+
+	/// The identifier of this migration. Which should be globally unique.
+	fn id() -> Self::Identifier {
+		MigrationId { pallet_id: *PALLET_MIGRATIONS_ID, version_from: 1, version_to: 2 }
+	}
+
+	/// The actual logic of the migration.
+	///
+	/// This function is called repeatedly until it returns `Ok(None)`, indicating that the
+	/// migration is complete. Ideally, the migration should be designed in such a way that each
+	/// step consumes as much weight as possible. However, this is simplified to perform one stored
+	/// value mutation per block.
+	fn step(
+		mut cursor: Option<Self::Cursor>,
+		meter: &mut WeightMeter,
+	) -> Result<Option<Self::Cursor>, SteppedMigrationError> {
+		let required = W::step();
+		// If there is not enough weight for a single step, return an error. This case can be
+		// problematic if it is the first migration that ran in this block. But there is nothing
+		// that we can do about it here.
+		if meter.remaining().any_lt(required) {
+			return Err(SteppedMigrationError::InsufficientWeight { required });
+		}
+
+		// We loop here to do as much progress as possible per step.
+		loop {
+			if meter.try_consume(required).is_err() {
+				break;
+			}
+
+			let mut iter = if let Some(last_key) = cursor {
+				// If a cursor is provided, start iterating from the stored value
+				// corresponding to the last key processed in the previous step.
+				// Note that this only works if the old and the new map use the same way to hash
+				// storage keys.
+
+				v1::Multisigs::<T>::iter_from(v1::Multisigs::<T>::hashed_key_for(
+					last_key.0, last_key.1,
+				))
+			} else {
+				// If no cursor is provided, start iterating from the beginning.
+				v1::Multisigs::<T>::iter()
+			};
+
+			// If there's a next item in the iterator, perform the migration.
+			if let Some((last_key1, last_key2, value)) = iter.next() {
+				// Migrate the `when` field (`Timepoint<BlockNumber>`) -> `maybe_when` (`Option<Timepoint<BlockNumber>>`)
+				let new_multisig = Multisig {
+					maybe_when: Some(value.when),
+					deposit: value.deposit,
+					depositor: value.depositor,
+					approvals: value.approvals,
+				};
+
+				// We can just insert here since the old and the new map share the same key-space.
+				// Otherwise it would have to invert the concat hash function and re-hash it.
+				Multisigs::<T>::insert(last_key1, last_key2, new_multisig);
+				cursor = Some((last_key1, last_key2)) // Return the processed key as the new cursor.
+			} else {
+				cursor = None; // Signal that the migration is complete (no more items to process).
+				break;
+			}
+		}
+		Ok(cursor)
```

Comment:
any update on this @liamaharon ?

Reply:
StorageVersion::new(2).put::<crate::Pallet<T>>();

Reply:
@liamaharon the problem I'm facing is I don't know where to put `StorageVersion::new(2).put::<crate::Pallet>();`. It should be put as the last step of the migration (if migration succeeded). 

- I couldn't find a relevant function in the trait `SteppedMigration`. 
- Putting it in `step()` function seems counter intuitive because it will be called multiple times, and may not be finalized.
- tried to track down the initial caller of `step()` function for the migration, but failed to do so. See my effort below in the details.

<details>

`SteppedMigration::step()` <- `SteppedMigration::transactional_step()` <- `SteppedMigrations::nth_transactional_step()` <- `exec_migration()` <- `progress_mbms()` <- `MultiStepMigrator::step()` <- ???

Should I also implement `MultiStepMigrator` for the pallet? Even then, which code piece is actually triggering the migration part? I'm a bit confused at this point.

</details>

Reply:
Can you put it immediately before you set the cursor to None?

Reply:
> Should I also implement MultiStepMigrator for the pallet? Even then, which code piece is actually triggering the migration part? I'm a bit confused at this point.

Not required

Reply:
> Can you put the new storage version immediately before you set the cursor to None?
yeah makes sense 🤦 that place makes sure migration is finished, thanks for pointing it out! I was looking for it in the wrong places